src/write_open_telemetry.cc

```diff
@@ -0,0 +1,275 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Authors:
+ *   Florian octo Forster <octo at collectd.org>
+ **/
+
+/* write_open_telemetry plugin configuration example
+ *
+ * <Plugin write_open_telemetry>
+ *   <Node "name">
+ *     Host "localhost"
+ *     Port "8080"
+ *     Path "/v1/metrics"
+ *   </Node>
+ * </Plugin>
+ */
+
+extern "C" {
+#include "collectd.h"
+
+#include "plugin.h"
+#include "utils/common/common.h"
+
+#include "utils/resource_metrics/resource_metrics.h"
+#include "utils/strbuf/strbuf.h"
+#include "utils_complain.h"
+
+#include <netdb.h>
+}
+
+#include <grpc++/grpc++.h>
+
+#include "opentelemetry/proto/collector/metrics/v1/metrics_service.grpc.pb.h"
+#include "utils/format_open_telemetry/format_open_telemetry.h"
+
+#ifndef OT_DEFAULT_HOST
+#define OT_DEFAULT_HOST "localhost"
+#endif
+
+#ifndef OT_DEFAULT_PORT
+#define OT_DEFAULT_PORT "4317"
+#endif
+
+using opentelemetry::proto::collector::metrics::v1::
+    ExportMetricsServiceResponse;
+using opentelemetry::proto::collector::metrics::v1::MetricsService;
+
+/*
+ * Private variables
+ */
+typedef struct {
+  char *name;
+  int reference_count;
+
+  char *host;
+  char *port;
+
+  resource_metrics_set_t resource_metrics;
+  cdtime_t staged_time;
+
+  std::unique_ptr<MetricsService::Stub> stub;
+
+  pthread_mutex_t mu;
+} ot_callback_t;
+
+static int export_metrics(ot_callback_t *cb) {
+  if (cb->stub == NULL) {
+    strbuf_t buf = STRBUF_CREATE;
+    strbuf_printf(&buf, "%s:%s", cb->host, cb->port);
+
+    auto chan =
+        grpc::CreateChannel(buf.ptr, grpc::InsecureChannelCredentials());
+    cb->stub = MetricsService::NewStub(chan);
+
+    STRBUF_DESTROY(buf);
+  }
+
+  auto req = format_open_telemetry_export_metrics_service_request(
+      cb->resource_metrics);
+
+  grpc::ClientContext context;
+  ExportMetricsServiceResponse resp;
+  grpc::Status status = cb->stub->Export(&context, *req, &resp);
+
+  if (!status.ok()) {
+    ERROR("write_open_telemetry plugin: Exporting metrics failed: %s",
+          status.error_message().c_str());
+    return -1;
+  }
+
+  if (resp.has_partial_success() &&
+      resp.partial_success().rejected_data_points() > 0) {
+    auto ps = resp.partial_success();
+    NOTICE("write_open_telemetry plugin: %" PRId64
+           " data points were rejected: %s",
+           ps.rejected_data_points(), ps.error_message().c_str());
+  }
+
+  delete req;
+  return 0;
+}
+
+/* NOTE: You must hold cb->send_lock when calling this function! */
+static int ot_flush_nolock(cdtime_t timeout, ot_callback_t *cb) {
+  if (cb->resource_metrics.num == 0) {
+    cb->staged_time = cdtime();
+    return 0;
+  }
+
+  /* timeout == 0  => flush unconditionally */
+  if (timeout > 0) {
+    cdtime_t now = cdtime();
+    if ((cb->staged_time + timeout) > now)
+      return 0;
```

Comment:
I think this would be easier to read with args reversed:
```suggestion
    if (now < (cb->staged_time + timeout))
      return 0;
```